File: mcp/src/velum_mcp/tools/inventory.py
"""MCP tools — Inventory CRUD."""

from __future__ import annotations

from typing import Any, Literal

from velum_mcp.client import get_client

InventoryType = Literal["static", "file", "dynamic-azure", "dynamic-aws", "dynamic-gcp", "terraform-workspace"]


async def list_inventory(project_id: int) -> list[dict[str, Any]]:
    """List all inventories in a project."""
    client = get_client()
    return await client.get(f"/projects/{project_id}/inventories") or []


async def get_inventory(project_id: int, inventory_id: int) -> dict[str, Any]:
    """Get details of a specific inventory."""
    client = get_client()
    return await client.get(f"/projects/{project_id}/inventories/{inventory_id}")


async def create_inventory(
    project_id: int,
    name: str,
    inventory_type: InventoryType = "static",
    inventory: str = "",
    ssh_key_id: int | None = None,
    become_key_id: int | None = None,
) -> dict[str, Any]:
    """Create a new inventory.

    Args:
        project_id: Target project.
        name: Inventory display name.
        inventory_type: Type — 'static' (YAML/INI text), 'file' (path),
                        'dynamic-aws', 'dynamic-azure', 'dynamic-gcp',
                        or 'terraform-workspace'.
        inventory: Inventory content (for static) or path/config (for others).
        ssh_key_id: SSH key to use for connecting to hosts.
        become_key_id: Become (sudo) key ID (optional).
    """
    client = get_client()
    body: dict[str, Any] = {
        "project_id": project_id,
        "name": name,
        "type": inventory_type,
        "inventory": inventory,
    }
    if ssh_key_id is not None:
        body["ssh_key_id"] = ssh_key_id
    if become_key_id is not None:
        body["become_key_id"] = become_key_id
    return await client.post(f"/projects/{project_id}/inventories", body)


async def update_inventory(
    project_id: int,
    inventory_id: int,
    name: str | None = None,
    inventory: str | None = None,
) -> dict[str, Any]:
    """Update an inventory's name or content."""
    client = get_client()
    current = await client.get(f"/projects/{project_id}/inventories/{inventory_id}")
    body = dict(current)
    if name is not None:
        body["name"] = name
    if inventory is not None:
        body["inventory"] = inventory
    return await client.put(f"/projects/{project_id}/inventories/{inventory_id}", body)


async def delete_inventory(project_id: int, inventory_id: int) -> str:
    """Delete an inventory."""
    client = get_client()
    await client.delete(f"/projects/{project_id}/inventories/{inventory_id}")
    return f"Inventory {inventory_id} deleted from project {project_id}."

File: mcp/src/tools/inventory.rs
use crate::client::VelumClient;
use crate::protocol::{prop_int, prop_int_opt, prop_str, prop_str_opt, Tool, ToolResult};
use serde_json::{json, Value};
use anyhow::Result;

pub fn definitions() -> Vec<Tool> {
    vec![
        Tool { name: "list_inventory".into(), description: "List all inventories in a project.".into(),
            input_schema: json!({"type":"object","properties":{"project_id":prop_int("Project ID")},"required":["project_id"]}) },
        Tool { name: "get_inventory".into(), description: "Get details of a specific inventory.".into(),
            input_schema: json!({"type":"object","properties":{"project_id":prop_int("Project ID"),"inventory_id":prop_int("Inventory ID")},"required":["project_id","inventory_id"]}) },
        Tool { name: "create_inventory".into(), description: "Create a new inventory (static YAML/INI, dynamic, terraform-workspace, or file-based).".into(),
            input_schema: json!({"type":"object","properties":{
                "project_id":prop_int("Project ID"),"name":prop_str("Display name"),
                "type":{"type":"string","enum":["static","file","dynamic-azure","dynamic-aws","dynamic-gcp","terraform-workspace"]},
                "inventory":prop_str_opt("Inventory content or path"),"ssh_key_id":prop_int_opt("SSH key ID")},"required":["project_id","name"]}) },
        Tool { name: "update_inventory".into(), description: "Update an inventory's name or content.".into(),
            input_schema: json!({"type":"object","properties":{"project_id":prop_int("Project ID"),"inventory_id":prop_int("Inventory ID"),"name":prop_str_opt("New name"),"inventory":prop_str_opt("New content")},"required":["project_id","inventory_id"]}) },
        Tool { name: "delete_inventory".into(), description: "Delete an inventory.".into(),
            input_schema: json!({"type":"object","properties":{"project_id":prop_int("Project ID"),"inventory_id":prop_int("Inventory ID")},"required":["project_id","inventory_id"]}) },
    ]
}

pub async fn call(name: &str, args: &Value, client: &VelumClient) -> Result<ToolResult> {
    let pid = args.get("project_id").and_then(Value::as_i64).unwrap_or_default();
    match name {
        "list_inventory" => Ok(ToolResult::json(&client.get(&format!("/projects/{pid}/inventories")).await?)),
        "get_inventory" => {
            let id = args["inventory_id"].as_i64().unwrap_or_default();
            Ok(ToolResult::json(&client.get(&format!("/projects/{pid}/inventories/{id}")).await?))
        }
        "create_inventory" => {
            let body = json!({"project_id":pid,"name":args["name"],"type":args.get("type").and_then(Value::as_str).unwrap_or("static"),
                "inventory":args.get("inventory").and_then(Value::as_str).unwrap_or(""),"ssh_key_id":args.get("ssh_key_id")});
            Ok(ToolResult::json(&client.post(&format!("/projects/{pid}/inventories"), &body).await?))
        }
        "update_inventory" => {
            let id = args["inventory_id"].as_i64().unwrap_or_default();
            let mut cur = client.get(&format!("/projects/{pid}/inventories/{id}")).await?;
            if let Some(v) = args.get("name").and_then(Value::as_str) { cur["name"] = json!(v); }
            if let Some(v) = args.get("inventory").and_then(Value::as_str) { cur["inventory"] = json!(v); }
            Ok(ToolResult::json(&client.put(&format!("/projects/{pid}/inventories/{id}"), &cur).await?))
        }
        "delete_inventory" => {
            let id = args["inventory_id"].as_i64().unwrap_or_default();
            client.delete(&format!("/projects/{pid}/inventories/{id}")).await?;
            Ok(ToolResult::text(format!("Inventory {id} deleted.")))
        }
        _ => Ok(ToolResult::error(format!("Unknown tool: {name}"))),
    }
}
